packages/arb-validator/rollupvalidator/txtrackerdb.go

```diff
@@ -0,0 +1,269 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package rollupvalidator
+
+import (
+	"encoding/binary"
+	"errors"
+
+	"github.com/hashicorp/golang-lru"
+
+	"github.com/offchainlabs/arbitrum/packages/arb-util/common"
+	"github.com/offchainlabs/arbitrum/packages/arb-util/machine"
+	"github.com/offchainlabs/arbitrum/packages/arb-validator/ckptcontext"
+	"github.com/offchainlabs/arbitrum/packages/arb-validator/structures"
+	"google.golang.org/protobuf/proto"
+)
+
+var txRecordPrefix = []byte{53}
+var nodeRecordPrefix = []byte{54}
+
+type nodeRecordKey struct {
+	height uint64
+	hash   common.Hash
+}
+
+func txRecordKey(txHash common.Hash) []byte {
+	return append(append([]byte{}, txRecordPrefix...), txHash.Bytes()...)
+}
+
+func nodeMetadataKey(key nodeRecordKey) []byte {
+	encodedHeight := make([]byte, 8)
+	binary.BigEndian.PutUint64(encodedHeight, key.height)
+
+	rawKey := append([]byte{}, nodeRecordPrefix...)
+	rawKey = append(rawKey, encodedHeight...)
+	return append(rawKey, key.hash.Bytes()...)
+}
+
+func newNodeMetadata(node *nodeInfo) *NodeMetadata {
+	return &NodeMetadata{LogBloom: node.calculateBloomFilter().Bytes()}
+}
+
+type txDB struct {
+	db                 machine.CheckpointStorage
+	confirmedNodeStore machine.NodeStore
+
+	confirmedNodeCache *lru.Cache
+
+	transactions     map[common.Hash]*TxRecord
+	nodeInfo         map[nodeRecordKey]*nodeInfo
+	nodeMetadata     map[nodeRecordKey]*NodeMetadata
+	nodeHashLookup   map[common.Hash]uint64
+	nodeHeightLookup map[uint64]common.Hash
+	chainAddress     common.Address
+}
+
+func newTxDB(db machine.CheckpointStorage, ns machine.NodeStore, chainAddress common.Address) (*txDB, error) {
+	lruCache, err := lru.New(500)
+	if err != nil {
+		return nil, err
+	}
+	return &txDB{
+		db:                 db,
+		confirmedNodeStore: ns,
+		confirmedNodeCache: lruCache,
+		transactions:       make(map[common.Hash]*TxRecord),
+		nodeInfo:           make(map[nodeRecordKey]*nodeInfo),
+		nodeMetadata:       make(map[nodeRecordKey]*NodeMetadata),
+		nodeHashLookup:     make(map[common.Hash]uint64),
+		nodeHeightLookup:   make(map[uint64]common.Hash),
+		chainAddress:       chainAddress,
+	}, nil
+}
+
+func (txdb *txDB) removeUnconfirmedNode(nodeHeight uint64) error {
+	nodeHash, err := txdb.lookupNodeHash(nodeHeight)
+	if err != nil {
+		return err
+	}
+	node, err := txdb.lookupNodeRecord(nodeHeight, nodeHash)
+	if err != nil {
+		return err
+	}
+	txdb.deleteNode(node)
+	return nil
+}
+
+func (txdb *txDB) deleteNode(node *nodeInfo) {
+	for _, txHash := range node.TransactionHashes {
+		delete(txdb.transactions, txHash)
+	}
+	delete(txdb.nodeHeightLookup, node.NodeHeight)
```

Comment:
Seems like there are race conditions here, unless protected by some external locking. Deletions here might race against activities that set the same values for the post-reorg world.

Reply:
So the intention here is for the `txDB` to not be thread safe and to instead have the `txTracker` use a readwrite mutex lock when accessing it. On thing that might help would be to comment on each method what type of lock it needs.

Reply:
Added comments to make this more clear